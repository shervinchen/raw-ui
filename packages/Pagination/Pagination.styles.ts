import css from 'styled-jsx/css';
import { useTheme } from '../Theme';

export const usePaginationItemStyles = () => {
  const theme = useTheme();

  return css.resolve`
    .raw-pagination-item {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      white-space: nowrap;
      min-width: 32px;
      height: 32px;
      padding: 0 8px;
      border: 1px solid ${theme.palette.neutral['200']};
      border-radius: 6px;
      font-size: 14px;
      color: ${theme.palette.neutral['600']};
      background-color: ${theme.palette.background};
      transition-property: border-color, background, color;
      transition-duration: 0.2s;
      transition-timing-function: ease;
      user-select: none;
      text-decoration: none;
      cursor: pointer;
    }
    .raw-pagination-item.raw-pagination-item-disabled {
      background-color: ${theme.palette.neutral['200']};
      color: ${theme.palette.neutral['400']};
      cursor: not-allowed;
    }
    .raw-pagination-item.raw-pagination-item-active {
      background-color: ${theme.palette.foreground};
      border-color: ${theme.palette.foreground};
      color: ${theme.palette.background};
    }
    .raw-pagination-item:not(.raw-pagination-item-active):not(
        .raw-pagination-item-disabled
      ):hover {
      border-color: ${theme.palette.foreground};
      color: ${theme.palette.foreground};
    }
    .raw-pagination-item:not(.raw-pagination-item-active):not(
        .raw-pagination-item-disabled
      ):active {
      background-color: ${theme.palette.neutral['200']};
    }
  `;
};
