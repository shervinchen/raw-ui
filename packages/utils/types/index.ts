export type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends Record<string, any> ? DeepPartial<T[P]> : T[P];
};

export type Merge<T, P> = P & Omit<T, keyof P>;
