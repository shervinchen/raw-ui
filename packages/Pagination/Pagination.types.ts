import { ComponentPropsWithoutRef } from 'react';
import { Merge } from '../utils';

export interface BasePaginationProps {
  page?: number;
  defaultPage?: number;
  count?: number;
  limit?: number;
  className?: string;
  onChange?: (page: number) => void;
}

export type PaginationProps = Merge<
  ComponentPropsWithoutRef<'nav'>,
  BasePaginationProps
>;

export interface PaginationPagesProps {
  pages: number[];
  internalPage: number;
  handlePageNumberClick: (page: number) => void;
}
