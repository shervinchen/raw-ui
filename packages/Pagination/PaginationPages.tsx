import React, { FC } from 'react';
import { MoreHorizontal } from 'react-feather';
import classNames from 'classnames';
import { PaginationPagesProps } from './Pagination.types';
import { useTheme } from '../Theme';
import { usePaginationItemStyles } from './Pagination.styles';

const PaginationPages: FC<PaginationPagesProps> = ({
  pages,
  internalPage,
  handlePageNumberClick,
}) => {
  const theme = useTheme();
  const { className: paginationItemClassName, styles: paginationItemStyles } =
    usePaginationItemStyles();

  return (
    <>
      {pages.map((pageItem, index) => {
        const isCurrentPage = pageItem === internalPage;
        return (
          <li key={index}>
            {pageItem !== 0 ? (
              <a
                className={classNames(
                  'raw-pagination-item',
                  isCurrentPage && 'raw-pagination-item-active',
                  paginationItemClassName
                )}
                aria-current={isCurrentPage ? 'page' : undefined}
                onClick={() => handlePageNumberClick(pageItem)}
              >
                {pageItem}
              </a>
            ) : (
              <span className="raw-pagination-ellipsis">
                <MoreHorizontal size={16} color={theme.palette.accents7} />
              </span>
            )}
          </li>
        );
      })}
      {paginationItemStyles}
      <style jsx>
        {`
          .raw-pagination-ellipsis {
            display: inline-flex;
            align-items: center;
            justify-content: center;
            width: 40px;
            height: 40px;
          }
        `}
      </style>
    </>
  );
};

export default PaginationPages;
