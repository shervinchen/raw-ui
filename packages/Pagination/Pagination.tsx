import React, { FC, useMemo } from 'react';
import { ChevronLeft, ChevronRight } from 'react-feather';
import classNames from 'classnames';
import { PaginationProps } from './Pagination.types';
import { useControlled } from '../utils/hooks';
import PaginationPages from './PaginationPages';
import { usePaginationItemStyles } from './Pagination.styles';

const range = (start: number, end: number) => {
  const length = end - start + 1;
  return Array.from({ length }, (_, i) => start + i);
};

const Pagination: FC<PaginationProps> = ({
  page,
  defaultPage = 1,
  count = 1,
  limit = 10,
  onChange,
  className = '',
  ...restProps
}) => {
  const [internalPage, setInternalPage] = useControlled<number>({
    defaultValue: defaultPage,
    value: page,
  });
  const classes = classNames('raw-pagination', className);
  const { className: paginationItemClassName, styles: paginationItemStyles } =
    usePaginationItemStyles();
  const pageCount = Math.ceil(count / limit);
  const boundaryCount = 5;
  const startPages = range(1, boundaryCount);
  const endPages = range(pageCount - boundaryCount + 1, pageCount);

  const pages: number[] = useMemo(() => {
    return [
      ...new Set(
        [
          ...(internalPage < boundaryCount ? startPages : [1]),
          internalPage,
          internalPage - 1,
          internalPage + 1,
          ...(internalPage > pageCount - boundaryCount + 1
            ? endPages
            : [pageCount]),
        ]
          .filter((value) => value >= 1 && value <= pageCount)
          .sort((a, b) => a - b)
      ),
    ].reduce<number[]>((previousValue, currentValue, currentIndex, array) => {
      previousValue.push(currentValue);
      if (
        array[currentIndex + 1] &&
        array[currentIndex + 1] - array[currentIndex] > 1
      ) {
        previousValue.push(0);
      }
      return previousValue;
    }, []);
  }, [internalPage, pageCount, startPages, endPages]);

  const clickPageNumberHandler = (pageNumber: number) => {
    setInternalPage(pageNumber);
    onChange?.(pageNumber);
  };

  const clickPreviousPageHandler = () => {
    if (internalPage === 1) {
      return;
    }
    clickPageNumberHandler(internalPage - 1);
  };

  const clickNextPageHandler = () => {
    if (internalPage === pageCount) {
      return;
    }
    clickPageNumberHandler(internalPage + 1);
  };

  return (
    <nav
      role="navigation"
      aria-label="pagination"
      className={classes}
      {...restProps}
    >
      <ul className="raw-pagination-list">
        <li>
          <a
            className={classNames(
              'raw-pagination-item',
              internalPage === 1 && 'raw-pagination-item-disabled',
              paginationItemClassName
            )}
            aria-label="Go to previous page"
            onClick={clickPreviousPageHandler}
          >
            <ChevronLeft size={16} />
          </a>
        </li>
        <PaginationPages
          pages={pages}
          internalPage={internalPage}
          clickPageNumberHandler={clickPageNumberHandler}
        />
        <li>
          <a
            className={classNames(
              'raw-pagination-item',
              internalPage === pageCount && 'raw-pagination-item-disabled',
              paginationItemClassName
            )}
            aria-label="Go to next page"
            onClick={clickNextPageHandler}
          >
            <ChevronRight size={16} />
          </a>
        </li>
      </ul>
      {paginationItemStyles}
      <style jsx>
        {`
          .raw-pagination-list {
            display: flex;
            align-items: center;
            gap: 8px;
            list-style: none;
            padding: 0;
            margin: 0;
          }
        `}
      </style>
    </nav>
  );
};

export default Pagination;
