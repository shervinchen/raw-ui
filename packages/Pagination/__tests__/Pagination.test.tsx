import React, { useState } from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Pagination from '../Pagination';
import { PaginationProps } from '../Pagination.types';

const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });

describe('Pagination', () => {
  test('should match the snapshot', () => {
    const { asFragment } = render(<Pagination />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('should support custom class name', () => {
    render(<Pagination className="custom-pagination" />);
    expect(screen.getByRole('navigation')).toHaveClass('custom-pagination');
  });

  test('should support default value', () => {
    render(<Pagination defaultPage={2} count={100} />);
    expect(screen.getByText('2')).toHaveClass('raw-pagination-item-active');
  });

  test('should support controlled value', async () => {
    const onChange = jest.fn();

    const Component = (props: PaginationProps) => {
      const [page, setPage] = useState(1);

      return (
        <Pagination
          page={page}
          count={100}
          onChange={(value) => {
            setPage(value);
            props.onChange?.(value);
          }}
        />
      );
    };

    render(<Component onChange={onChange} />);
    const pageItemSecond = screen.getByText('2');
    await user.click(pageItemSecond);
    expect(pageItemSecond).toHaveClass('raw-pagination-item-active');
    expect(onChange).toHaveBeenCalledWith(2);
  });

  test('should support go to previous when page is not first', async () => {
    render(<Pagination defaultPage={2} count={100} />);
    const pageItemPrevious = screen.getByLabelText('Go to previous page');
    const pageItemFirst = screen.getByText('1');
    const pageItemSecond = screen.getByText('2');
    expect(pageItemSecond).toHaveClass('raw-pagination-item-active');
    await user.click(pageItemPrevious);
    expect(pageItemFirst).toHaveClass('raw-pagination-item-active');
    expect(pageItemSecond).not.toHaveClass('raw-pagination-item-active');
  });

  test('should not support go to previous when page is first', async () => {
    render(<Pagination defaultPage={1} count={100} />);
    const pageItemPrevious = screen.getByLabelText('Go to previous page');
    const pageItemFirst = screen.getByText('1');
    expect(pageItemFirst).toHaveClass('raw-pagination-item-active');
    await user.click(pageItemPrevious);
    expect(pageItemFirst).toHaveClass('raw-pagination-item-active');
  });

  test('should support go to next when page is not last', async () => {
    render(<Pagination defaultPage={2} count={100} />);
    const pageItemNext = screen.getByLabelText('Go to next page');
    const pageItemSecond = screen.getByText('2');
    const pageItemThird = screen.getByText('3');
    expect(pageItemSecond).toHaveClass('raw-pagination-item-active');
    await user.click(pageItemNext);
    expect(pageItemThird).toHaveClass('raw-pagination-item-active');
    expect(pageItemSecond).not.toHaveClass('raw-pagination-item-active');
  });

  test('should not support go to next when page is last', async () => {
    render(<Pagination defaultPage={10} count={100} />);
    const pageItemNext = screen.getByLabelText('Go to next page');
    const pageItemLast = screen.getByText('10');
    expect(pageItemLast).toHaveClass('raw-pagination-item-active');
    await user.click(pageItemNext);
    expect(pageItemLast).toHaveClass('raw-pagination-item-active');
  });
});
