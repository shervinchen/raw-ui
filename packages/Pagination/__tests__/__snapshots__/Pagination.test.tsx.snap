// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should match the snapshot 1`] = `
<DocumentFragment>
  <nav
    aria-label="pagination"
    class="jsx-4082017107 raw-pagination"
    role="navigation"
  >
    <ul
      class="jsx-4082017107 raw-pagination-list"
    >
      <li
        class="jsx-4082017107"
      >
        <a
          aria-label="Go to previous page"
          class="jsx-4082017107 raw-pagination-item raw-pagination-item-disabled jsx-2619780212"
        >
          <svg
            fill="none"
            height="16"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="15 18 9 12 15 6"
            />
          </svg>
        </a>
      </li>
      <li
        class="jsx-3237814599"
      >
        <a
          aria-current="page"
          class="jsx-3237814599 raw-pagination-item raw-pagination-item-active jsx-2619780212"
        >
          1
        </a>
      </li>
      <li
        class="jsx-4082017107"
      >
        <a
          aria-label="Go to next page"
          class="jsx-4082017107 raw-pagination-item raw-pagination-item-disabled jsx-2619780212"
        >
          <svg
            fill="none"
            height="16"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <polyline
              points="9 18 15 12 9 6"
            />
          </svg>
        </a>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
